cpp/src/lists/contains.cu

```diff
@@ -105,48 +140,168 @@ struct search_list_fn {
-   *         forward or reverse order.
-   * @param size The number of elements in the list.
-   * @return A pair of {begin, end} iterators to iterate through the range `[0, size)`.
-   */
+ private:
   template <bool forward>
-  static __device__ auto element_index_pair_iter(size_type const size)
+  __device__ inline size_type search_list(list_device_view list) const
   {
-    if constexpr (forward) {
-      return thrust::pair(thrust::make_counting_iterator(0), thrust::make_counting_iterator(size));
+    using cudf::experimental::row::lhs_index_type;
+    using cudf::experimental::row::rhs_index_type;
+
+    auto const [begin, end] = element_index_pair_iter<forward>(list.size());
+    auto const found_iter =
+      thrust::find_if(thrust::seq, begin, end, [&] __device__(auto const idx) {
+        return !list.is_null(idx) &&
+               d_comp(static_cast<lhs_index_type>(list.element_offset(idx)),
+                      static_cast<rhs_index_type>(search_key_is_scalar ? 0 : list.row_index()));
+      });
+    // If the key is found, return its found position in the list from `found_iter`.
+    return found_iter == end ? NOT_FOUND_SENTINEL : *found_iter;
+  }
+};
+
+/**
+ * @brief Function to search for key element(s) in the corresponding rows of a lists column,
+ * specialized for non-nested types.
+ */
+template <bool search_key_is_scalar,
+          typename Element,
+          typename InputIterator,
+          typename SearchKeyType>
+void index_of_non_nested(InputIterator input_it,
+                         size_type num_rows,
+                         size_type* output_it,
+                         SearchKeyType const& search_keys,
+                         bool search_keys_have_nulls,
+                         duplicate_find_option find_option,
+                         rmm::cuda_stream_view stream)
+{
+  auto const do_search = [&](auto const keys_iter) {
+    thrust::transform(rmm::exec_policy(stream),
+                      input_it,
+                      input_it + num_rows,
+                      keys_iter,
+                      output_it,
+                      search_list_non_nested_fn{find_option});
+  };
+
+  if constexpr (search_key_is_scalar) {
+    auto const keys_iter = cudf::detail::make_optional_iterator<Element>(
+      search_keys, nullate::DYNAMIC{search_keys_have_nulls});
+    do_search(keys_iter);
+  } else {
+    auto const keys_cdv_ptr = column_device_view::create(search_keys, stream);
+    auto const keys_iter    = cudf::detail::make_optional_iterator<Element>(
+      *keys_cdv_ptr, nullate::DYNAMIC{search_keys_have_nulls});
+    do_search(keys_iter);
+  }
+}
+
+/**
+ * @brief Function to search for index of key element(s) in the corresponding rows of a lists
+ * column, specialized for nested types.
+ */
+template <bool search_key_is_scalar, typename InputIterator, typename SearchKeyType>
+void index_of_nested_types(InputIterator input_it,
```

Comment:
The naming here is asymmetric. Why `index_of_nested_types` and `index_of_non_nested`? Both (or neither) should use the suffix `_types`.